---
title: "Genome-wide DNA binding in HEPG2 (2kb window)"
author: "EM"
date: "4/18/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
library(Gviz)
library(ggpubr)
library(ggplot2)
library(tidyverse)
library(GenomicRanges)
source("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/util/intersect_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/util/plotting_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/util/_setup.R")
```

# Goal:

Here we aim to download all available DNA binding protein (DBP) profiles in a single cell state (measured by ChIP-seq) . This will allow us to investigate the binding properties of hundreds of DBPs in the same cellular context or background. We aim to address several questions: (i) What are the number of peaks and genome coverage for each DBP? (ii) What are the binding preferences for promoters, gene-bodies and intergenic genomic regions? (iii) What are the similarities and differences across DBPs based on their genome-wide binding profiles genome-wide? (iv) What properties or preferences do promoters have for binding events. (iv) Are there reservoir promoters in HepG2 as defined in k562 previously? (v) How does binding to a promoter affect the transcriptional output of that promoter?

To address these questions we have curated a set of X,000 ChIPs-eq data sets comprised of 486 DBPs in HEPG2 cells from the ENCODE consortrium. We required duplicate ChIP-seq experiments for a given DBP and other criterion that can be found here :

<https://www.encodeproject.org/report/?type=Experiment&status=released&assay_slims=DNA+binding&biosample_ontology.term_name=HepG2&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell+line&files.read_length=100&files.read_length=76&files.read_length=75&files.read_length=36&assay_title=Control+ChIP-seq&assay_title=Histone+ChIP-seq&files.run_type=single-ended>

## These samples were selected on the following criteria:

1)  "chromatin" interaction data, then DNA binding data, cell line HEPG2, "TF-Chip-seq".
2)  We further selected "TF Chip-seq", "Control chip-seq" and "Histone Chip-seq".
3)  We selected several read lengths to get the most DNA binding proteins (DBPs)
4)  Read lengths: 100, 76, 75, 36
5)  ONLY SINGLE END READS (this eliminates 54 samples)

### Experimental data was downloading by (ENCODE report.tsv):

<https://www.encodeproject.org/report.tsv?type=Experiment&status=released&assay_slims=DNA+binding&biosample_ontology.term_name=HepG2&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell+line&files.read_length=100&files.read_length=76&files.read_length=75&files.read_length=36&assay_title=Control+ChIP-seq&assay_title=Histone+ChIP-seq&files.run_type=single-ended>

### The FASTQ files were downloaded with:

"<https://www.encodeproject.org/metadata/?status=released&assay_slims=DNA+binding&biosample_ontology.term_name=HepG2&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell+line&files.read_length=100&files.read_length=76&files.read_length=75&files.read_length=36&assay_title=Control+ChIP-seq&assay_title=Histone+ChIP-seq&files.run_type=single-ended&type=Experiment>"

MD5sums were checked with all passing (see encode_file_info function to reterive MD5Sum values that are not available from the encode portal (/util)

### Processing data:

We processed all the read alignments and peak calling using the NF_CORE ChIP-seq pipeline: (nfcore/chipseq v1.2.1)

## Next we created consensus peaks that overlap in both replicates

Our strategy was to take peaks in each replicate and find all overlapping peak windows. We then took the union length of the overlapping range in each peak window.

```{r creating-consensus-peaks}
# create_consensus_peaks requires an annotation .GTF file - loading in Gencode v32 annotations.
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2022/data/genomes/gencode.v32.annotation.gtf")
# Creating consensus peaks function to create a .bed file of overlapping peaks in each replicate.
# /util/intersect_functions.R
# TODO run this only on final knit
# create_consensus_peaks <- create_consensus_peaks(broadpeakfilepath = "/scratch/Shares/rinnclass/CLASS_2022/data/test_work/all_peak_files")
# exporting consensus peaks .bed files
# TODO run this only on final knit
# for(i in 1:length(consensus_peaks)) {
#  rtracklayer::export(consensus_peaks[[i]],
#                     paste0("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/class_exeRcises/Analysis/11_r_functions/consensus_peaks",
#                             names(consensus_peaks)[i],
#                             "_consensus_peaks.bed"))
# }
```

# loading in consensus peaks to prevent rerunning create_consensus_peaks function

```{r consensus-peak-filtering}
# Loading in files via listing and rtracklayer import
consensus_fl <- list.files("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/class_exeRcises/Analysis/11_r_functions/consensus_peaks", full.names = T)
# importing (takes ~5min)
consensus_peaks <- lapply(consensus_fl, rtracklayer::import)
# cleaning up file names
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/class_exeRcises/Analysis/11_r_functions/consensus_peaks/|_consensus_peaks.bed","", consensus_fl)
# Filtering consensus peaks to those DBPs with at least 250 peaks
num_peaks_threshold <- 250
num_peaks <- sapply(consensus_peaks, length)
filtered_consensus_peaks <- consensus_peaks[num_peaks > num_peaks_threshold]
# Result: these were the DBPs that were filtered out.
filtered_dbps <- consensus_peaks[num_peaks < num_peaks_threshold]
names(filtered_dbps)
# We have this many remaining DBPs
length(filtered_consensus_peaks)
```

## Now we will determine the peak number and genome coverage for each DBP.

```{r peak number and coverage per DBP}
# Let's start with loading in the number of peaks each DBP has -- using length.
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_peaks),
                           "num_peaks" = sapply(filtered_consensus_peaks, length))
# total genomic coverage of peaks for each dbp
num_peaks_df$total_peak_length <- sapply(filtered_consensus_peaks, function(x) sum(width(x)))
# Plotting distribution of peak number per dbp
hist(num_peaks_df$total_peak_length)
```

# Now we will create promoter annotations for lncRNA and mRNA and both.

We have created a funciton get_promter_regions that has up and downstream parameters

```{r promoter-gene-body-annotations}
# creating lncRNA and mRNA promoters
big_number = 1
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"), upstream = big_number, downstream = big_number)
names(lncrna_mrna_promoters) <- lncrna_mrna_promoters$gene_id
rtracklayer::export(lncrna_mrna_promoters, "analysis/results/lncRNA_mrna_promoters.gtf")
# creating lncRNAs promoter
lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA", upstream = big_number, downstream = big_number) 
names(lncrna_promoters) <- lncrna_promoters$gene_id
rtracklayer::export(lncrna_promoters, "analysis/results/lncRNA_promoters.gtf")
# creating mRNA promoters
mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding", upstream = big_number, downstream = big_number)
names(mrna_promoters) <- mrna_promoters$gene_id
rtracklayer::export(lncrna_promoters, "analysis/results/mRNA_promoters.gtf")
# creating all genebody annotation
lncrna_mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                     gencode_gr$gene_type %in% c("lncRNA", "protein_coding")]
names(lncrna_mrna_genebody) <- lncrna_mrna_genebody$gene_id
rtracklayer::export(lncrna_mrna_genebody, "analysis/results/lncrna_mrna_genebody.gtf")
# creating lncRNA genebody annotation
lncrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                                gencode_gr$gene_type %in% c("lncRNA")]
names(lncrna_genebody) <- lncrna_genebody$gene_id
rtracklayer::export(lncrna_mrna_genebody, "analysis/results/lncrna_genebody.gtf")
# creating mRNA genebody annotation
mrna_genebody <- gencode_gr[gencode_gr$type == "gene" & 
                              gencode_gr$gene_type %in% c("protein_coding")]
names(mrna_genebody) <-mrna_genebody$gene_id
rtracklayer::export(lncrna_mrna_genebody, "analysis/results/mrna_genebody.gtf")
```

# Determining the overlaps of chip peaks with promoters and genebodys

```{r DBP CHIP peaks overlap with promoters and genebody }
# creating index to subset lncRNA and mRNA annotations
lncrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "lncRNA"]
mrna_gene_ids <- lncrna_mrna_genebody$gene_id[lncrna_mrna_genebody$gene_type == "protein_coding"]
# using count peaks per feature returns number of annotation overlaps for a given DBP (takes ~5min)
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, type = "counts")
# adding data to num_peaks_df
num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)
num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])
# gene body overlaps 
genebody_peak_counts <- count_peaks_per_feature(lncrna_mrna_genebody, 
                                                filtered_consensus_peaks, 
                                                type = "counts")
# adding data to num_peaks_df
num_peaks_df$peaks_overlapping_genebody <- rowSums(genebody_peak_counts)
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])
num_peaks_df$peaks_overlapping_mrna_genebody <- rowSums(genebody_peak_counts[,mrna_gene_ids])
write_csv(num_peaks_df, "analysis/results/num_peaks_df.csv")
```

# Plotting peak annotation features for DBPs

```{r plotting peak annotation features}
num_peaks_df <- read_csv("analysis/results/num_peaks_df.csv")
# Distriubtion of peak numbers of all 460 DBPs
ggplot(num_peaks_df, aes(x = num_peaks)) + 
  geom_histogram(bins = 70)
# Plotting number of peaks versus total genome coverage
ggplot(num_peaks_df, aes(x = num_peaks, y = total_peak_length)) +
  geom_point() + 
  geom_smooth(method = "gam", se = TRUE, color = "black", lty = 2)+
         
  ylab("BP covered") +
  xlab("Number of peaks") +
  ggtitle("Peak count vs. total bases covered")
ggsave("analysis/figures/peak_num_vs_coverage.pdf")
# Plotting number of peaks versus peaks overlapping promoters
ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_promoters)) +
  xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping promoters") +
  ggtitle("Relationship Between Number of DBP Peaks and Promoter Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=FALSE, formula = 'y ~ x',
              color = "#a8404c") +
  stat_regline_equation(label.x = 35000, label.y = 18000) +
  ylim(0,60100) +
  xlim(0,60100)
ggsave("analysis/figures/3_peak_num_vs_promoter_coverage.pdf")
# Plotting peak overlaps with genebody
ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_genebody)) +
  xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping genes") +
  ggtitle("Relationship Between Number of DBP Peaks and Gene Body Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=F, formula = 'y ~ x',
              color = "#a8404c") +
  stat_regline_equation(label.x = 35000, label.y = 18000) +
  ylim(0,60100) +
  xlim(0,60100)
ggsave("analysis/figures/4_peak_num_vs_gene_body_coverage.pdf")
# there is a large amount of data explained (almost all by genebodys)
# Let's see what percentage of the genome genebodys cover:
reduced_gene_bodies <- gencode_gr[gencode_gr$type == "gene"] %>%
  GenomicRanges::reduce() %>%
  width() %>%
  sum()
# percentage of gene bodies in genome
reduced_gene_bodies/3.2e9
```

# Counting the number of overlaps at each promoter

promoters are the cols and DBPs rows thus we can retrieve the number of binding events at each promoter unlike the "counts parameter" that just gives total number of overlaps

```{r peak occurence matrix}
# Creating matrix of promoters(annotation feature) as cols and DBPs as rows (takes ~5min)
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, 
                                               type = "occurrence")
# test to make sure everything is in right order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))
# Formatting final data.frame from peak occurence matrix
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "1_kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))
# exporting
write_csv(peak_occurence_df, "analysis/results/peak_occurence_dataframe.csv")
```

# WHERE DO YOU WANT TO GO FROM HERE ?

# Bimodal promoter plot
```{r}
samplesheet <- read_rds("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/19_rnaseq/final_samplesheet.rds")

salmon_tpm <- read.csv("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/class_exeRcises/Analysis/18_RNAseq/results/salmon/salmon_merged_gene_tpm.csv")
# since we have replicates we will take the mean TPM for each fractionation.
# while we are at it we will add more intuitive row and col names 
# finally we are making sure the TPM table is in same order as samplesheet
tpm <- salmon_tpm %>% 
  # pivot longer here is going to take cols 2 through the rest (the samples conditions)
  # So each gene will have 10 rows with the sample_id = experiment name/condition
  # TPM will be TPM 
  pivot_longer(cols = 2:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  # This will merge the sample sheet by sample_id column
  merge(samplesheet) %>%
  # we group the data by gene_id condition thus there are 2 values in 5 condition
  group_by(gene_id, condition) %>%
  # taking the mean by using summarize function (this is the mean of 2 res in each condition)
  summarize(tpm = mean(tpm, na.rm = T)) %>%
  # bringing the table back to "normal" with pivot wider
  # the condition will be come the cols with names_from = with a prefix tpm_condition name
  # So we are back to 5 columns since each condition has only one mean value now
  # Those TPM mean values are populating the rows now
  pivot_wider(names_from = condition, values_from = tpm, names_prefix = "tpm_")



promoter_features_df <- read.csv("analysis/results/peak_occurence_dataframe.csv")
# let's take a look and refresh our memories of this data frame we created of ChiP binding properties
View(promoter_features_df)
# promoter_features_df <- read.csv("/scratch/Shares/rinnclass/CLASS_2022/JR/CLASS_2022/class_exeRcises/analysis/12_peak_features/peak_occurence_dataframe.csv")
# Now we can merge in the TPM data to this data.frame
# This will merge on the on teh gene_id column
promoter_features_df <- merge(promoter_features_df, tpm)
# saving this file
write.csv(promoter_features_df, "promoter_feature_df_tpm.csv")
# let's plot the number of DBPs bound as a densuty plot.
promoter_features_df_3kb <- read.csv("/scratch/Shares/rinnclass/CLASS_2022/EM/CLASS_2022/class_exeRcises/Analysis/20_binding/results/promoter_features_df.csv")

ggplot() +
  geom_histogram(aes(x = promoter_features_df[promoter_features_df$number_of_dbp > 0,]$number_of_dbp), fill = "Red", alpha = 0.5, bins = 100)+
  geom_histogram(aes(x = promoter_features_df_3kb[promoter_features_df_3kb$number_of_dbp > 0,]$number_of_dbp), fill = "Blue", alpha = 0.5, bins = 100)
# Cool we see the bimodal distribution again.
# Let's save
  ggsave("analysis/figures/DBP_binding_density_plot.pdf")
```


# Bimodal plot by biotype
```{r}
mrna_promoter_peak_counts <- count_peaks_per_feature(mrna_promoters, filtered_consensus_peaks, type = "counts")

lncrna_promoter_peak_counts <- count_peaks_per_feature(lncrna_promoters, filtered_consensus_peaks, type = "counts")

#mrna_promoter_features_df <- merge(mrna_promoter_peak_counts, tpm)

mrna_dbp_counts <- colSums(mrna_promoter_peak_counts)

#lncrna_promoter_features_df <- merge(lncrna_promoter_peak_counts, tpm)

lncrna_dbp_counts <- colSums(lncrna_promoter_peak_counts)

ggplot() +
  geom_density(aes(x = mrna_dbp_counts[mrna_dbp_counts>0]), color = "Blue") + 
  geom_density(aes(x = lncrna_dbp_counts[lncrna_dbp_counts>0]), color = "Red")
```


# Binding versus expression
```{r}

```



1. How does the binding vs. expression change for the high binding promoters (> ~300 DBPs) vs low binders? And does this trend change across subcellular fractions?

```{r}

#num_peaks_df_new <- num_peaks_df %>%
 # mutate(high_binding = ())



```

1) Reduce
2) Find overlaps
3) 

